### ISSUE TITLE
[ENH]: Handling of closed streamlines in streamplot

### ISSUE THREAD

**Issue by quenot** [matplotlib/matplotlib#27261]
### Problem

With the current streamplot implementation, when asking for unbroken streamlines with given start points on fields with theoretically closed streamlines, the integration process spirals in practice until the maximum trajectory length is reached. The issue is more detailed, with an example and a workaround in [this stack overflow question and answer](https://stackoverflow.com/questions/77344931/how-can-i-avoid-spiraling-streamlines-in-matplotlib-pyplot-streamplot). When it is theoretically known that some streamlines should loop over themselves, we would like to be able to specify it and integrate the streamlines so that they are actually closed or, at least, stop the integration when a "quasi-loop" has been detected, by passing again close enough to a start point and/or when a 360 degrees rotation has been observed for the velocity vector.

### Proposed solution

A minimal and quite easy solution for single-direction integration could be to keep track of the accumulated rotation of the velocity vector along the trajectory, to continuously check the distance between the starting point and additional segments and to stop the integration at the closest point to the starting point after about one complete plus or minus 360 degrees velocity vector rotation and if the closest point is close enough to the starting point. Something similar though a bit more complicated could be done for two-direction integration.

A second-level solution would be to do exactly the same but while additionally modifying "on the fly" the original vector field with a very small rotation so that, after the integration, "the closest point to the starting point" becomes as close as possible to the starting point. The optimal "very small rotation" angle could be theoretically predicted and/or iteratively determined. Though this may appear as tampering with the source data, if there are good theoretical reasons to believe that the streamlines should be closed (the user should know what she or he is doing when using the "force_closed_streamlines" option), this is overall more likely to correct accumulated integration errors than to distort the result.

Having looked at the code, it seems to me that I should be able to implement this and maybe I will do so anyway for my personal use as the workaround that I am currently using is very impractical. However, it would certainly be done much better by matplotlib developers. Please let me know if this feature if of interest, if some matplotlib developers would consider implementing it, and if yes, how long it could take.

**Comment by quenot** [comment]
> If I wanted closed streamlines, I'd calculate the stream function and contour that. In fact if I was sure that the vector field was non-divergent, I'd always do that anyways rather than using `streamplot`. Making the end of a streamline exactly hit the starting point of a streamline is going to be very difficult and any fuzzy logic is going to be incorrect for some vector fields.

I would indeed do that if I had a direct access to the potential function from which the field is derived. There theoretically exists one for the fields I am considering but I do not have access to it and I don't know either how to reconstruct it from the field itself. Therefore, I don't know what to contour. Making the end of a streamline "exactly" hit the starting point of a streamline is a bit complicated to implement but a priori doable. I don't see what fuzzy logic has to do with fields nor what the issue would be.

**Comment by jklymak** [comment]
Because the streamplot is just an approximation, it is very unlikely the streamlines start and ends will exactly match up numerically, so you would need to have a tolerance to say a streamline is "touching" itself and call it closed.  

You empirically calculate the stream function from the velocity field by psi = -du/dy and/or dv/dx and integrate using known boundary conditions (usually that a boundary is a streamline).   The resulting stream function is also just an approximation, but subsequently calling `contour` will give you closed (approximate) contours.  

Perhaps this should be an example, but I don't think there is a way to add it to streamplot.

**Comment by quenot** [comment]
> You empirically calculate the stream function from the velocity field by psi = -du/dy and/or dv/dx and integrate using known boundary conditions (usually that a boundary is a streamline). The resulting stream function is also just an approximation, but subsequently calling `contour` will give you closed (approximate) contours.

Great! I learned something today, thanks a lot! Actually, it is u = -dpsi/dy and v dpsi/dx but I got the idea and the code is just;
```
dx, dy = x[0,1]-x[0,0], y[1,0]-y[0,0]
psi =np.cumsum(u,axis=0)*dy-np.cumsum(v[0:1,:],axis=1)*dx
plt.contour(x, y, psi, levels = 31)
```
It seems to work well with magnetic fields generated using magpylib. Fixing the starting points should be easy.
With that, it seems pointless to implement a "close_streamline" feature in streamplot, at least for me.

**Comment by jklymak** [comment]
Lets reopen as a documentation request, as I think a nice example along these lines would be very helpful to point folks towards...

**Comment by quenot** [comment]
Sure, it would have saved me time if I had encountered this before. I will also make the suggestion on the magpylib side.

**Comment by quenot** [comment]
Hoping this helps, I have provided an  [example with magpylib on stack overflow](https://stackoverflow.com/a/77419388/13707125).

**Comment by quenot** [comment]
@story645 Sure, I am preparing [something a bit better finished and complete for magpylib](https://github.com/magpylib/magpylib/discussions/677). I will be happy to share it on the matplotlib side too. It may take a few days. I'll let you know when it is ready.

**Comment by jklymak** [comment]
Just a heads up that we are very unlikely to accept magpylib as a documentation dependency.

**Comment by quenot** [comment]
@jklymak I expected that and it makes sense. I will just have to use another source field. Everything else should not change.

**Comment by anntzer** [comment]
As a side point, I also think your SO post may be showing a real bug wrt. masking? (the streamlines that cross the masked region due to a too-large step size?)

**Comment by quenot** [comment]
@anntzer Yes, I feel that this is a bug as I mentioned in my SO post. Actually, I would not only expect the streamlines not to go though masked region but I would also expect them to reach the boundary of the masked region when possible by some kind of Euler step there too. Having looked at the code, both do not seem easy to implement.

Reducing the step size seems to be a reasonable and easy workaround even if it slows down the computation (checking for masked region crossing would be costly too anyway). It seems that the only way to control the step size is by increasing the density as the step size is bounded so that the step cannot go through more than one density cell at once. Maybe a direct control of the step size as an additional parameter to streamplot would be more convenient (there is a hard default value at 0.1 in the code). This parameter might also be used to control the accuracy of the integration in complement of the grid resolution.

**Comment by quenot** [comment]
@jklymak @story645
Hello.
Here are some news. Regarding the stream function approach, it turned out to be quite unstable and/or complicated to make work as expected, especially in the case of magnets as this is discussed in https://github.com/magpylib/magpylib/issues/678#issuecomment-1803421634

**Comment by quenot** [comment]
@jklymak @story645 @anntzer 
Hello.
Here are some news. Regarding the stream function approach, unfortunately, it turned out to be quite unstable and/or complicated to make work as expected, especially in the case of magnets, as this is discussed in https://github.com/magpylib/magpylib/issues/678#issuecomment-1803421634
However, I finally implemented, with the help of @Alexboiboi the solution that I initially proposed above: https://github.com/magpylib/magpylib/issues/678#issuecomment-1807193641. It currently relies on pyvista, which is a dependency that I guess you are also very unlikely to accept. It is like that because I built upon a partial implementation from @Alexboiboi which was based on it and also because matplotlib's streamplot does not currently return the streamlines without plotting them and modifying them after they have been returned does not act back on the plot but the "hard part" does not rely itself on pyvista.
I believe that it can be quite easily integrated directly in matplotlib's streamplot as a "close_streamlines" option if you think that this can be of interest for matplotlib users.
It has not been intensively tested but it successfully worked in two hard cases: degenerated (theoretically reduced to one point) streamlines and streamlines with (theoretically) abrupt changes in direction. With non-singular streamlines, the looping is already very good and the distortion is very small, almost unnecessary. For what I got, pyvista default's and matplotlib's integration methods are in principle the same.
